cs.1
.\" Copyright (c) 2014 Andy Kosela.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"   notice, this list of conditions and the following disclaimer in the
.\"   documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.Dd July 19, 2017
.Dt CS 1
.Os
.Sh NAME
.Nm cs
.Nd concurrent ssh client
.Sh SYNOPSIS
.Nm cs
.Bk -words
.Op Fl emqrstuVv1
.Op Fl c Ar file
.Op Fl cmd
.Op Fl cname
.Op Fl d Ar file
.Op Fl dd
.Op Fl du Ar path
.Op Fl f Ar script.sh
.Op Fl h Ar hosts_file
.Op Fl hba
.Op Fl i Ar identity_file
.Op Fl io
.Op Fl ip
.Op Fl l Ar login_name
.Op Fl lip
.Op Fl mx
.Op Fl nc
.Op Fl nmap
.Op Fl ns
.Op Fl o Ar output_file
.Op Fl P Ar port
.Op Fl p Ar path
.Op Fl ping
.Op Fl soa
.Op Fl to Ar timeout
.Op Fl top
.Op Fl tr
.Op Fl tri
.Op Fl uname
.Op Fl vm
.Op Ar command
.Sm off
.Oo
.Op Ar user No @
.Ar host
.Oc
.Sm on
.Ar ...
.Sh DESCRIPTION
.Nm
is a program for concurrently executing local or remote commands on
multiple hosts.
It is using OpenSSH for running remote commands.
Its main purpose is to help automate and manage large network of hosts.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl c Ar file
Copy
.Ar file
to the remote machine.
.It Fl cmd
Run an arbitrary local command concurrently on multiple hosts.
.It Fl cname
Run a local DNS query of type CNAME.
.It Fl d Ar file
Download
.Ar file
from the remote machine.
It will be saved in a directory named after the remote host only when
you download from multiple servers.
.It Fl dd
Prints basic hardware information for a specific host (sudo(8) and
dmidecode(8) required).
.It Fl du Ar path
Prints the list of largest files in a specific directory (sudo(8)
required, units in M).
.It Fl e
Prints hosts with errors only.
.It Fl f Ar script.sh
Run a local shell script on the remote host.
.It Fl h Ar hosts_file
Read hostnames from the given
.Ar hosts_file .
Lines in the
.Ar hosts_file
can include commented lines beginning with a
.Sq Li #
and only one host per line is allowed.
.It Fl hba
Lists installed HBAs and their WWPNs (Linux only).
.It Fl i Ar identity_file
Selects a file from which the identity (private key) for public key
authentication is read.
The default is
.Ar ~/.ssh/id_rsa .
.It Fl io
Run remote iostat(1) on a specific host.
.It Fl ip
Lists network interfaces.
.It Fl l Ar login_name
Specifies the user to log in as on the remote machine.
This also may be specified on a per-host basis on the command line.
.It Fl lip
Scans storage interconnects for new luns (Linux only).
.It Fl m
Lists mounted filesystems on a speficic host.
.It Fl mx
Run a local DNS query of type MX.
.It Fl nc
Tests a specific port with nc(1).
Default is 22/tcp.
.It Fl nmap
Scans host with nmap(1).
.It Fl ns
Run a local DNS query of type NS.
.It Fl o Ar output_file
Saves standard output and standard error to a file.
.It Fl P Ar port
Port to connect to on the remote host.
.It Fl p Ar path
Specifies remote or local path for files in a remote copy or download
mode.
.It Fl ping
Sends ICMP ECHO_REQUEST to network hosts.
.It Fl q
Quiet mode.
Supresses verbose standard output from remote machines.
This mode reports success or failure only.
.It Fl r
Recursively copy entire directories.
It follows symbolic links encountered in the tree traversal.
.It Fl s
Sort output.
.It Fl soa
Run a local DNS query of type SOA.
.It Fl t
Force pseudo-tty allocation.
.It Fl to Ar timeout
Specifies the timeout (in seconds) used when connecting to the SSH
server.
The default value is 4 seconds.
.It Fl top
Run remote top(1) in batch mode on a specific host.
.It Fl tr
Run local traceroute(8).
.It Fl tri
Run local traceroute(8) using ICMP (local sudo(8) required).
.It Fl uname
Prints remote system information including OS version.
.It Fl u
Run remote uptime(1) on a specific host.
.It Fl V
Displays the version number and exit.
.It Fl v
Verbose mode.
Causes
.Nm
to print debugging messages from ssh(1) about its progress.
This is helpful in debugging connection, authentication, and
configuration problems.
Multiple
.Fl v
options increase the verbosity.
The maximum is 3.
.It Fl vm
Run remote vmstat(8) on a specific host.
.It Fl 1
One line mode, useful for sorting output later.
.El
.Sh AUTHENTICATION
The default method for authentication is a public key authentication
which serves its purpose when dealing with multiple hosts.
You can read more about public key authentication in
.Xr ssh 1 .
.Sh EXIT STATUS
.Ex -std cs
.Sh EXAMPLES
Run a series of commands on hosts foo and bar:
.Pp
.Dl $ cs 'uptime; uname -a' foo bar
.Pp
Run a command on multiple hosts specified in a hosts_file:
.Pp
.Dl $ cs -h hosts_file uptime
.Pp
Run a local shell script on multiple hosts:
.Pp
.Dl $ cs -f script.sh foo{1..100}
.Pp
Run a local shell script with sudo(8) on multiple hosts:
.Pp
.Dl $ cs -t -f script.sh foo{1..100}
.Pp
Copy file to multiple hosts using a specific remote path:
.Pp
.Dl $ cs -c file -h hosts_file -p /foo/bar
.Pp
Download file from host:~ foo to a current working directory:
.Pp
.Dl $ cs -d file foo
.Pp
Download recursively files from /foo/bar from multiple hosts to a
specific local path /tmp with subdirectories named after remote hosts:
.Pp
.Dl $ cs -r -d '/foo/bar/*' -h hosts_file -p /tmp
.Pp
Run a command on multiple hosts and sort the output:
.Pp
.Dl $ cs -1 -h hosts_file 'free -m | grep Swap' | sort -rnk4 | head
.Pp
Run local ping(1) on multiple hosts:
.Pp
.Dl $ cs -ping foo{1..100}
.Pp
Show name and size of all files matching the glob pattern:
.Pp
.Dl $ cs -du '/var/log/mes*' foo{1..100}
.Sh SEE ALSO
.Xr scp 1 ,
.Xr ssh 1 ,
.Xr ssh_config 5 ,
.Xr sudo 8
.Sh AUTHORS
.An Andy Kosela Aq akosela@andykosela.com


CHANGES
go-cs v0.9						    Jul 19, 2017

  * Bugfix: improve running sudo(8).
  * Change: update man page.
  * Feature: add '-hba, -io, -lip, -m' options.

go-cs v0.8						    Jul 10, 2017

  * Bugfix: fix '-ping' on OpenBSD and NetBSD.
  * Bugfix: fix bugs in '-du, -top, -vm'.
  * Change: separate '-ip' from '-dd'.
  * Change: update man page.

go-cs v0.7						    Jun 30, 2017

  * Change: refactor some functions.
  * Feature: add '-cmd, -cname, -dd, -du, -e, -mx, -nc, -nmap, -ns,
    -ping, -soa, -top, -tr, -tri, -u, -uname, -vm' options.

go-cs v0.6						    Oct 10, 2016

  * Bugfix: new '-t' option to make sudo(8) happy.
  * Change: rename '-t' to '-to'.

go-cs v0.5						    May 11, 2015

  * Bugfix: clean checkArgs() function.
  * Bugfix: fix '-d' when downloading same file from multiple servers.
    Now it creates subdirectories named after remote hosts.

go-cs v0.4						    Apr 23, 2015

  * Change: '-c', '-d', '-f' now accept arguments.
  * Feature: update and expand man page.

go-cs v0.3						    Dec 30, 2014

  * Change: rename '-v' to '-V'.
  * Feature: add '-v' option to display verbose mode.
  * Feature: add '-1' option to display one line mode.  It is especially
    helpful for sorting output later.

go-cs v0.2						   Sept 29, 2014

  * Bugfix: fix bug in error routine.
  * Bugfix: clean run() function.
  * Bugfix: fix bug with login option.
  * Change: rename cs.go -> cs/cs.go.
  * Feature: add godoc comments.
  * Feature: automatically add host key to known_hosts file.
  * Feature: add '-v' option to display version.
  * Security: securely call ssh(1) by hardcoding its path.

go-cs v0.1 					    	    Sept 8, 2014

  * The first public version.


cs/cs.go
// Copyright 2014 Andy Kosela.  All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

// Concurrent ssh client.
// cs is a program for concurrently executing local or remote commands
// on multiple hosts.  It is using OpenSSH for running remote commands.
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"os/exec"
	"runtime"
	"sort"
	"strings"
	"time"
)

const (
	du      = "/usr/bin/du"
	host    = "/usr/bin/host"
	hostn	= "/bin/hostname"
	ifconf 	= "/sbin/ifconfig"
	lspci	= "/sbin/lspci"
	mpath	= "/sbin/multipath"
	nc      = "/usr/bin/nc"
	scp     = "/usr/bin/scp"
	ssh     = "/usr/bin/ssh"
	top     = "/usr/bin/top"
	uptime  = "/usr/bin/uptime"
	timeFmt = "02-Jan-2006 15:04:05"
)

func createFile(path string) *os.File {
	file, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY|os.O_APPEND,
		0666)
	if err != nil {
		fmt.Println("cs:", err)
		os.Exit(1)
	}
	return file
}

func openFile(path string) *os.File {
	file, err := os.Open(path)
	if err != nil {
		fmt.Println("cs:", err)
		os.Exit(1)
	}
	return file
}

func readFile(file *os.File) []string {
	var s []string
	r := bufio.NewReader(file)
	for {
		line, _ := r.ReadString('\n')
		if strings.HasPrefix(line, "#") {
			continue
		} else if line == "" {
			break
		}
		s = append(s, line)
	}
	defer file.Close()
	return s
}

func checkArgs(nocmd int, copy, download, file, hostsfile string, tty *bool,
	argv []string) (string, []string) {
	var command string
	var hosts []string

	if nocmd == 1 {
		if hostsfile == "" {
			if len(argv) < 1 {
				flag.Usage()
			}
			command = ""
			hosts = argv[0:]
		} else {
			f := openFile(hostsfile)
			hosts = readFile(f)
			command = ""
		}

		return command, hosts
	}

	if file != "" || copy != "" || download != "" {
		if hostsfile == "" {
			if len(argv) < 1 {
				flag.Usage()
			}
			hosts = argv[0:]
		} else {
			f := openFile(hostsfile)
			hosts = readFile(f)
		}
		if file != "" {
			f := openFile(file)
			if *tty {
				c := strings.Join(readFile(f), "")
				command = "sudo sh <<'EOF'" + c + "EOF"
			} else {
				command = strings.Join(readFile(f), "")
			}
		}
	} else {
		if hostsfile == "" {
			if len(argv) < 2 {
				flag.Usage()
			}
			command = argv[0]
			hosts = argv[1:]
		} else {
			f := openFile(hostsfile)
			hosts = readFile(f)
			command = argv[0]
		}
	}

	return command, hosts
}

func exist(hostname, path string) string {
	if fi, err := os.Stat(path + "/" + hostname); err != nil {
		os.Mkdir(path+"/"+hostname, 0755)
		return path + "/" + hostname
	} else if fi.IsDir() {
		return path + "/" + hostname
	}
	os.Mkdir(path+"/"+hostname+".host", 0755)
	return path + "/" + hostname + ".host"
}

func run(command, hostname, id, login, path, port, timeout, copy, disku,
	download string, dd, cname, hba, io, ip, lcmd, lip, netcat, nmap, ns,
	mount, mx, one, png, recursive, soa, up, verbose1, verbose2, verbose3,
	top1, tr, tri, tty, uname, vm *bool, ddir int, f *os.File) string {

	hostname = strings.Trim(hostname, "\n")
	batchmode := "-oBatchMode=yes"
	strict := "-oStrictHostKeyChecking=no"
	tout := "-oConnectTimeout=" + timeout
	flag := "-"
	flag2 := ""
	if *verbose1 {
		flag = "-v"
	} else if *verbose2 {
		flag = "-vv"
	} else if *verbose3 {
		flag = "-vvv"
	}

	var cmd *exec.Cmd
	if copy != "" && *recursive {
		if login != "" {
			cmd = exec.Command(scp, flag+"r", "-i", id, "-P", port,
				batchmode, strict, tout, copy, login+"@"+
				hostname+":"+path)
		} else {
			cmd = exec.Command(scp, flag+"r", "-i", id, "-P", port,
				batchmode, strict, tout, copy, hostname+":"+
				path)
		}
	} else if copy != "" {
		if login != "" {
			cmd = exec.Command(scp, flag+"i", id, "-P", port,
				batchmode, strict, tout, copy, login+"@"+
				hostname+":"+path)
		} else {
			cmd = exec.Command(scp, flag+"i", id, "-P", port,
				batchmode, strict, tout, copy, hostname+":"+
				path)
		}
	} else if download != "" && *recursive {
		if ddir == 1 {
			path = exist(hostname, path)
		}

		if login != "" {
			cmd = exec.Command(scp, flag+"r", "-i", id, "-P", port,
				batchmode, strict, tout, login+"@"+hostname+
				":"+download, path)
		} else {
			cmd = exec.Command(scp, flag+"r", "-i", id, "-P", port,
				batchmode, strict, tout, hostname+":"+download,
				path)
		}
	} else if download != "" {
		if ddir == 1 {
			path = exist(hostname, path)
		}

		if login != "" {
			cmd = exec.Command(scp, flag+"i", id, "-P", port,
				batchmode, strict, tout, login+"@"+hostname+":"+
				download, path)
		} else {
			cmd = exec.Command(scp, flag+"i", id, "-P", port,
				batchmode, strict, tout, hostname+":"+download,
				path)
		}
	} else if *cname {
		cmd = exec.Command(host, "-tcname", hostname)
	} else if *dd {
		c := "sudo dmidecode -s system-product-name " +
			"; sudo dmidecode -s system-serial-number"
		if login != "" {
			cmd = exec.Command(ssh, "-q", flag+"tti", id, "-l",
				login, "-p", port, batchmode, strict, tout,
				hostname, c)
		} else {
			cmd = exec.Command(ssh, "-q", flag+"tti", id, "-p",
				port, batchmode, strict, tout, hostname, c)
		}
	} else if disku != "" {
		c := "if [ `uname -s` == OpenBSD ]; then sudo " + du +
			" -akx " + disku + " |sort -rn 2>/dev/null |head -20" +
			"; else sudo " + du + " -amx " + disku +
			" |sort -rn 2>/dev/null |head -20; fi"
		if login != "" {
			cmd = exec.Command(ssh, "-q", flag+"tti", id, "-l",
				login, "-p", port, batchmode, strict, tout,
				hostname, c)
		} else {
			cmd = exec.Command(ssh, "-q", flag+"tti", id, "-p",
				port, batchmode, strict, tout, hostname, c)
		}
	} else if *hba {
		c := "if [ `uname -s` == Linux ]; then " + lspci + " |" +
			"grep -i fibre; grep -v zZzZ -H " +
			"/sys/class/fc_host/host*/port_name; else echo " +
			"'cs: your OS is not supported'; fi"
		if login != "" {
			cmd = exec.Command(ssh, flag+"i", id, "-l", login, "-p",
				port, batchmode, strict, tout, hostname, c)
		} else {
			cmd = exec.Command(ssh, flag+"i", id, "-p", port,
				batchmode, strict, tout, hostname, c)
		}
	} else if *io {
		c := "if [ `uname -s` == Linux ]; then iostat -mx |grep Dev " +
			"|awk '{OFS=\"\t\" ; print $1, $4, $5, $6, $7, $10, " +
			"$12, $9}'; iostat -mx |grep -e dm- -e sd -e hd | " +
			"sort -rnk10 2>/dev/null |head -20 | " +
			"awk '{OFS=\"\t\" ; print $1, $4, $5, $6, $7, $10, " +
			"$12, $9}'; elif [ `uname -s` == FreeBSD ]; then " +
			"iostat -x | grep qlen ; iostat -x |grep -v device | " +
			"sort -rnk9 2>/dev/null |head -20; " +
			"else echo 'cs: your OS is not supported'; fi"
		if login != "" {
			cmd = exec.Command(ssh, flag+"i", id, "-l", login, "-p",
				port, batchmode, strict, tout, hostname, c)
		} else {
			cmd = exec.Command(ssh, flag+"i", id, "-p", port,
				batchmode, strict, tout, hostname, c)
		}
	} else if *ip {
		if login != "" {
			cmd = exec.Command(ssh, flag+"i", id, "-l", login, "-p",
				port, batchmode, strict, tout, hostname, ifconf)
		} else {
			cmd = exec.Command(ssh, flag+"i", id, "-p", port,
				batchmode, strict, tout, hostname, ifconf)
		}
	} else if *lip {
		c := "if [ `uname -s` == Linux ]; then for e in `ls " +
			"/sys/class/fc_host/host*/issue_lip`; do echo $e; " +
			"echo '1' >$e; done; for e in `ls " +
			"/sys/class/scsi_host/host*/scan`; do echo $e; echo " +
			"'- - -' >$e; done; echo '# multipath -F'; " + mpath +
			" -F; echo '# multipath'; " + mpath +
			"; else echo 'cs: your OS is not supported'; fi"
		command = "sudo sh <<'EOF'" + "\n" + c + "\n" + "EOF"
		if login != "" {
			cmd = exec.Command(ssh, "-q", flag+"tti", id, "-l",
				login, "-p", port, batchmode, strict, tout,
				hostname, command)
		} else {
			cmd = exec.Command(ssh, "-q", flag+"tti", id, "-p",
				port, batchmode, strict, tout, hostname,
				command)
		}
	} else if *mount {
		if login != "" {
			cmd = exec.Command(ssh, flag+"i", id, "-l", login, "-p",
				port, batchmode, strict, tout, hostname,
				"mount")
		} else {
			cmd = exec.Command(ssh, flag+"i", id, "-p", port,
				batchmode, strict, tout, hostname, "mount")
		}
	} else if *mx {
		cmd = exec.Command(host, "-tmx", hostname)
	} else if *ns {
		cmd = exec.Command(host, hostname)
	} else if *nmap {
		cmd = exec.Command("nmap", hostname)
	} else if *netcat {
		cmd = exec.Command(nc, "-w1", hostname, port)
	} else if *png {
		var c string
		if runtime.GOOS == "linux" {
			c = "ping -nc1 -s16 -W3 " + hostname + " |grep from"
		} else if runtime.GOOS == "freebsd" {
			c = "ping -nc1 -s16 -W3000 " + hostname + " |grep from"
		} else {
			c = "ping -nc1 -s16 -w3 " + hostname + " |grep from"
		}
		cmd = exec.Command("/bin/sh", "-c", c)
	} else if *lcmd {
		scmd := strings.Split(command, " ")

		switch len(scmd) {
		case 1:
			cmd = exec.Command(scmd[0], hostname)
		case 2:
			cmd = exec.Command(scmd[0], scmd[1], hostname)
		case 3:
			cmd = exec.Command(scmd[0], scmd[1], scmd[2], hostname)
		case 4:
			cmd = exec.Command(scmd[0], scmd[1], scmd[2], scmd[3],
				hostname)
		case 5:
			cmd = exec.Command(scmd[0], scmd[1], scmd[2], scmd[3],
				scmd[4], hostname)
		}
	} else if *soa {
		cmd = exec.Command(host, "-tsoa", hostname)
	} else if *top1 {
		c := "if [ `uname -s` == Linux ]; then " + top +
			" -cbn1 |grep -ve '\\[' -e " + top + " 2>/dev/null" +
			" |head -20; else " + top + " -b 13; fi"
		if login != "" {
			cmd = exec.Command(ssh, flag+"i", id, "-l", login, "-p",
				port, batchmode, strict, tout, hostname, c)
		} else {
			cmd = exec.Command(ssh, flag+"i", id, "-p", port,
				batchmode, strict, tout, hostname, c)
		}
	} else if *tr {
		cmd = exec.Command("traceroute", hostname)
	} else if *tri {
		cmd = exec.Command("sudo", "traceroute", "-I", hostname)
	} else if *uname {
		c := "if [ `uname -s` == Linux ]; then uname -a; " +
			"cat /etc/redhat-release; else uname -a; fi"
		if login != "" {
			cmd = exec.Command(ssh, flag+"i", id, "-l", login, "-p",
				port, batchmode, strict, tout, hostname, c)
		} else {
			cmd = exec.Command(ssh, flag+"i", id, "-p", port,
				batchmode, strict, tout, hostname, c)
		}
	} else if *up {
		if login != "" {
			cmd = exec.Command(ssh, flag+"i", id, "-l", login, "-p",
				port, batchmode, strict, tout, hostname, uptime)
		} else {
			cmd = exec.Command(ssh, flag+"i", id, "-p", port,
				batchmode, strict, tout, hostname, uptime)
		}
	} else if *vm {
		c := "if [ `uname -s` == Linux ]; then vmstat -SM; " +
			"elif [ `uname -s` == FreeBSD ]; then vmstat -h; " +
			"else vmstat; fi"
		if login != "" {
			cmd = exec.Command(ssh, flag+"i", id, "-l", login, "-p",
				port, batchmode, strict, tout, hostname, c)
		} else {
			cmd = exec.Command(ssh, flag+"i", id, "-p", port,
				batchmode, strict, tout, hostname, c)
		}
	} else {
		if *tty {
			flag2 = "tt"
		}
		if login != "" {
			cmd = exec.Command(ssh, "-q", flag+flag2+"i", id, "-l",
				login, "-p", port, batchmode, strict, tout,
				hostname, command)
		} else {
			cmd = exec.Command(ssh, "-q", flag+flag2+"i", id, "-p",
				port, batchmode, strict, tout, hostname,
				command)
		}
	}

	buf, err := cmd.CombinedOutput()
	if err != nil {
		return hostname + ": " + string(buf)
	}

	if *one {
		return hostname + ": " + string(buf)
	}
	return hostname + ":\n" + string(buf)
}

func main() {
	flag.Usage = func() {
		fmt.Println(
`usage: cs [-emqrstuVv1] [-c file] [-cmd] [-cname] [-d file] [-dd] [-du path]
	  [-f script.sh] [-h hosts_file] [-hba] [-i identity_file] [-io] [-ip]
	  [-l login_name] [-lip] [-mx] [-nc] [-nmap] [-ns] [-o output_file]
	  [-P port] [-p path] [-ping] [-soa] [-to timeout] [-top] [-tr]
	  [-tri] [-uname] [-vm] [command] [[user@]host] ...`)
		os.Exit(1)
	}

	lcmd := flag.Bool("cmd", false, "Local command")
	cname := flag.Bool("cname", false, "CNAME")
	copy := flag.String("c", "", "Copy")
	dd := flag.Bool("dd", false, "Dmidecode")
	disku := flag.String("du", "", "Du")
	download := flag.String("d", "", "Download")
	error := flag.Bool("e", false, "Error")
	file := flag.String("f", "", "Script file")
	hba := flag.Bool("hba", false, "HBA")
	hostsfile := flag.String("h", "", "Hosts file")
	id := flag.String("i", string(os.Getenv("HOME")+"/.ssh/id_rsa"),
		"Identity file")
	io := flag.Bool("io", false, "Iostat")
	ip := flag.Bool("ip", false, "IP")
	lip := flag.Bool("lip", false, "LIP")
	login := flag.String("l", "", "Login name")
	mount := flag.Bool("m", false, "Mounts")
	mx := flag.Bool("mx", false, "MX")
	netcat := flag.Bool("nc", false, "Netcat")
	nmap := flag.Bool("nmap", false, "Nmap")
	ns := flag.Bool("ns", false, "NS")
	one := flag.Bool("1", false, "One line")
	out := flag.String("o", "", "Output filename")
	path := flag.String("p", ".", "Path")
	png := flag.Bool("ping", false, "Ping")
	port := flag.String("P", "22", "SSH port")
	quiet := flag.Bool("q", false, "Quiet")
	recursive := flag.Bool("r", false, "Recursive")
	soa := flag.Bool("soa", false, "SOA")
	sorted := flag.Bool("s", false, "Sort")
	timeout := flag.String("to", "4", "Timeout")
	top1 := flag.Bool("top", false, "Top")
	tr := flag.Bool("tr", false, "Traceroute")
	tri := flag.Bool("tri", false, "Traceroute -I")
	tty := flag.Bool("t", false, "Force pseudo-tty allocation")
	uname := flag.Bool("uname", false, "Uname")
	up := flag.Bool("u", false, "Uptime")
	version := flag.Bool("V", false, "Version")
	verbose1 := flag.Bool("v", false, "Verbose mode 1")
	verbose2 := flag.Bool("vv", false, "Verbose mode 2")
	verbose3 := flag.Bool("vvv", false, "Verbose mode 3")
	vm := flag.Bool("vm", false, "Vmstat")
	flag.Parse()
	argv := flag.Args()

	if *version {
		fmt.Println("cs 0.9")
		os.Exit(1)
	}

	nocmd := 0
	if *cname || *dd || *disku != "" || *hba || *io || *ip || *lip ||
		*mount || *netcat || *nmap || *ns || *mx || *png || *soa ||
		*top1 || *tr || *tri || *uname || *up || *vm {
		nocmd = 1
	}

	command, hosts := checkArgs(nocmd, *copy, *download, *file, *hostsfile,
		tty, argv)

	ddir := 0
	if len(hosts) > 1 {
		ddir = 1
	}

	var f *os.File
	if *out != "" {
		now := time.Now()
		nowStr := now.Format(timeFmt)
		f = createFile(*out)
		f.WriteString("------ START: " + nowStr + " ------\n")
	}

	var mk, mk2 []string
	if *sorted {
		mk = make([]string, len(hosts))
		if *quiet {
			mk2 = make([]string, len(hosts))
		}
	}

	output := make(chan string)
	for _, hostname := range hosts {
		go func(hostname string) {
			output <- run(command, hostname, *id, *login, *path,
				*port, *timeout, *copy, *disku, *download, dd,
				cname, hba, io, ip, lcmd, lip, netcat, nmap, ns,
				mount, mx, one, png, recursive, soa, up,
				verbose1, verbose2, verbose3, top1, tr, tri,
				tty, uname, vm, ddir, f)
		}(hostname)
	}

	e, err := 0, 0
	for i := 0; i < len(hosts); i++ {
		select {
		case c := <-output:
			err = 0
			if !strings.Contains(c, ":\n") {
				e++
				err = 1
				if !strings.Contains(c, "\n") {
					c = c + "\n"
				}
			}

			if *sorted {
				if *error && err == 0 {
					continue
				}
				if *quiet {
					if *out != "" {
						mk[i] = c
					}
					split := strings.Split(c, ":")
					if err == 1 {
						mk2[i] = split[0] +"\t[ERROR]\n"
						continue
					}
					mk2[i] = split[0] + "\t[OK]\n"
					continue
				}
				mk[i] = c
				continue
			}
			if *out != "" {
				if *error && err == 0 {
					continue
				}
				f.WriteString(c)
			}
			if *quiet {
				split := strings.Split(c, ":")
				if *error && err == 0 {
					continue
				}

				if err == 1 {
					fmt.Println(split[0] + "\t[ERROR]")
					continue
				}
				fmt.Println(split[0] + "\t[OK]")
				continue
			}

			if *error && err == 0 {
				continue
			}
			fmt.Print(c)
		}
	}

	if *sorted {
		sort.Strings(mk)
		if *out != "" {
			for _, v := range mk {
				f.WriteString(v)
			}
		}
		if *quiet {
			sort.Strings(mk2)
			for _, v := range mk2 {
				fmt.Print(v)
			}
		} else {
			for _, v := range mk {
				fmt.Print(v)
			}
		}
	}

	if !*one {
		fmt.Printf("hosts = %d, errors = %d\n", len(hosts), e)
	}
	if *out != "" {
		now := time.Now()
		nowStr := now.Format(timeFmt)
		f.WriteString("------   END: " + nowStr + " ------\n")
		f.Close()
	}
}
